File: src/utils/jwt.ts
/**
 * JWT HS256 自行實現
 * 使用 Web Crypto API 的 HMAC-SHA256,無外部依賴
 */

export interface JwtClaims {
  /** 管理員 ID */
  sub: string;
  /** 用戶名 */
  username: string;
  /** 用戶代碼 (如 "10001") */
  ucode: string;
  /** 真實姓名 */
  realname: string;
  /** 角色代碼 (逗號分隔, 如 "R101,R102") */
  rcodes: string;
  /** 是否超級管理員 (ucode="10001" 跳過所有權限檢查) */
  isSuper: boolean;
  /** 權限列表 (resource:action 格式, 如 ["content:index", "content:add"]) */
  permissions: string[];
  /** 簽發時間 (Unix 時間戳, 秒) */
  iat: number;
  /** 過期時間 (Unix 時間戳, 秒) */
  exp: number;
  /** 唯一標識 (用於黑名單) */
  jti: string;
}

/** Base64URL 編碼 (無填充) */
function base64UrlEncode(data: ArrayBuffer | string): string {
  const bytes = typeof data === 'string' ? new TextEncoder().encode(data) : new Uint8Array(data);
  let binary = '';
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

/** Base64URL 解碼 */
function base64UrlDecode(str: string): Uint8Array {
  const padded = str.replace(/-/g, '+').replace(/_/g, '/');
  const binary = atob(padded);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

/** 使用 Web Crypto API 計算 HMAC-SHA256 簽名 */
async function hmacSha256(message: string, secret: string): Promise<ArrayBuffer> {
  const key = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode(secret),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign'],
  );
  return crypto.subtle.sign('HMAC', key, new TextEncoder().encode(message));
}

/** 簽發 JWT Token */
export async function signJwt(claims: JwtClaims, secret: string): Promise<string> {
  const header = JSON.stringify({ alg: 'HS256', typ: 'JWT' });
  const payload = JSON.stringify(claims);
  const headerB64 = base64UrlEncode(header);
  const payloadB64 = base64UrlEncode(payload);
  const signingInput = `${headerB64}.${payloadB64}`;
  const signature = await hmacSha256(signingInput, secret);
  const sigB64 = base64UrlEncode(signature);
  return `${signingInput}.${sigB64}`;
}

/** 驗證 JWT Token,返回 Claims */
export async function verifyJwt(token: string, secret: string): Promise<JwtClaims | null> {
  const parts = token.split('.');
  if (parts.length !== 3) return null;

  const signingInput = `${parts[0]}.${parts[1]}`;
  const expectedSig = base64UrlDecode(parts[2]);
  const actualSig = new Uint8Array(await hmacSha256(signingInput, secret));

  // 常量時間比較簽名
  if (expectedSig.length !== actualSig.length) return null;
  let diff = 0;
  for (let i = 0; i < expectedSig.length; i++) {
    diff |= expectedSig[i] ^ actualSig[i];
  }
  if (diff !== 0) return null;

  // 解析 payload
  const claimsBytes = base64UrlDecode(parts[1]);
  const claimsJson = new TextDecoder().decode(claimsBytes);
  const claims: JwtClaims = JSON.parse(claimsJson);

  // 檢查過期時間
  const now = Math.floor(Date.now() / 1000);
  if (claims.exp < now) return null;

  return claims;
}

/** 從 Authorization 頭提取 JWT Token */
export function extractToken(authorization: string | null | undefined): string | null {
  if (!authorization?.startsWith('Bearer ')) return null;
  return authorization.slice(7);
}

/** 生成唯一標識 (用於 jti) */
export function genUuid(): string {
  const now = Date.now();
  const rand = Math.random().toString(16).slice(2, 10);
  return `${now.toString(16)}${rand}`.padStart(32, '0');
}

File: src/util/jwt.rs
/// JWT HS256 自行實現
/// 使用 sha2 + hmac 純 Rust 實現,避免 jsonwebtoken 的 C 依賴
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use hmac::{Hmac, Mac};
use serde::{Deserialize, Serialize};
use sha2::Sha256;

type HmacSha256 = Hmac<Sha256>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JwtClaims {
    /// 管理員 ID
    pub sub: String,
    /// 用戶名
    pub username: String,
    /// 簽發時間 (Unix 時間戳, 秒)
    pub iat: u64,
    /// 過期時間 (Unix 時間戳, 秒)
    pub exp: u64,
    /// 唯一標識 (用於黑名單)
    pub jti: String,
}

/// 簽發 JWT Token
pub fn sign_jwt(claims: &JwtClaims, secret: &str) -> Result<String, String> {
    let header = r#"{"alg":"HS256","typ":"JWT"}"#;
    let header_b64 = URL_SAFE_NO_PAD.encode(header);
    let claims_json = serde_json::to_string(claims).map_err(|e| e.to_string())?;
    let claims_b64 = URL_SAFE_NO_PAD.encode(claims_json);

    let signing_input = format!("{}.{}", header_b64, claims_b64);

    let mut mac = HmacSha256::new_from_slice(secret.as_bytes()).map_err(|e| e.to_string())?;
    mac.update(signing_input.as_bytes());
    let signature = mac.finalize().into_bytes();
    let sig_b64 = URL_SAFE_NO_PAD.encode(signature);

    Ok(format!("{}.{}", signing_input, sig_b64))
}

/// 驗證 JWT Token,返回 Claims
pub fn verify_jwt(token: &str, secret: &str) -> Result<JwtClaims, String> {
    let parts: Vec<&str> = token.split('.').collect();
    if parts.len() != 3 {
        return Err("invalid token format".to_string());
    }

    let signing_input = format!("{}.{}", parts[0], parts[1]);
    let expected_sig = URL_SAFE_NO_PAD
        .decode(parts[2])
        .map_err(|e| e.to_string())?;

    let mut mac = HmacSha256::new_from_slice(secret.as_bytes()).map_err(|e| e.to_string())?;
    mac.update(signing_input.as_bytes());
    mac.verify_slice(&expected_sig)
        .map_err(|_| "invalid signature".to_string())?;

    let claims_bytes = URL_SAFE_NO_PAD
        .decode(parts[1])
        .map_err(|e| e.to_string())?;
    let claims: JwtClaims =
        serde_json::from_slice(&claims_bytes).map_err(|e| e.to_string())?;

    // 檢查過期時間
    let now = chrono::Utc::now().timestamp() as u64;
    if claims.exp < now {
        return Err("token expired".to_string());
    }

    Ok(claims)
}

/// 從請求頭提取 JWT Token
pub fn extract_token(authorization: &str) -> Option<&str> {
    if authorization.starts_with("Bearer ") {
        Some(&authorization[7..])
    } else {
        None
    }
}

/// 生成 UUID (簡易版,用 jti)
pub fn gen_uuid() -> String {
    let now = chrono::Utc::now().timestamp_millis() as u128;
    format!("{:032x}", now)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_jwt_sign_verify() {
        let claims = JwtClaims {
            sub: "1".to_string(),
            username: "admin".to_string(),
            iat: 1000,
            exp: 9999999999,
            jti: "test-jti".to_string(),
        };
        let secret = "test-secret";
        let token = sign_jwt(&claims, secret).unwrap();
        let verified = verify_jwt(&token, secret).unwrap();
        assert_eq!(verified.sub, "1");
        assert_eq!(verified.username, "admin");
    }

    #[test]
    fn test_jwt_invalid_token() {
        let result = verify_jwt("invalid.token.here", "secret");
        assert!(result.is_err());
    }
}
